tensorflow_recommenders_addons/dynamic_embedding/python/keras/layers/embedding.py

```diff
@@ -0,0 +1,459 @@
+                                    dtype=value_dtype)
+
+  def call(self, ids):
+    """
+    Make feature ids input becomes a embedding tensor. There are two stages
+    here. The first stage will lookup the dyamic_embedding.Variable and get
+    raw output of the embedding with shape (shape(ids), embedding_size).
+    The second stage will squash the raw embedding output to a fixed shape
+    embedding. Here the fixed shape is (batch_size, embedding_size). We
+    need a fixed shape here because the keras layer requires a explicitly
+    set shape in output.
+
+    It could be override in an inherited class from Embedding, to achieve
+    a custom embedding layer.
+
+    Args:
+      ids: feature ids of the input. It should be same dtype as the key_dtype
+        of the layer.
+
+    Returns:
+      A embedding output with shape (batch_size, embedding_size) if the input
+        are batched, else its shape is (embedding_size).
+    """
+    ids = tf.convert_to_tensor(ids)
+    input_shape = tf.shape(ids)
+    lookup_result = de.shadow_ops.embedding_lookup(self.shadow, ids)
+    lookup_result = tf.reshape(
+        lookup_result, tf.concat([input_shape, [self.embedding_size]], 0))
+    embedding = reduce_pooling(lookup_result, combiner=self.combiner)
+    return embedding
+
+  # TODO(Lifann) Need test cases.
+  def get_config(self):
+    _initializer = self.params.initializer
+    if _initializer is None:
+      _initializer = tf.keras.initializers.Zeros()
+    _max_norm = None
+    if isinstance(self.max_norm, tf.keras.constraints.Constraint):
+      _max_norm = tf.keras.constraints.serialize(self.max_norm)
+
+    if self.params.restrict_policy:
+      _restrict_policy = self.params.restrict_policy.__class__
+    else:
+      _restrict_policy = None
+
+    config = {
+        'embedding_size':
+            self.embedding_size,
+        'combiner':
+            self.combiner,
+        'key_dtype':
+            self.params.key_dtype,
+        'value_dtype':
+            self.params.value_dtype,
+        'initializer':
+            tf.keras.initializers.serialize(_initializer),
+        'devices':
+            self.params.devices if self.keep_distribution else None,
+        'name':
+            self.name,
+        'trainable':
+            self.trainable,
+        'bp_v2':
+            self.params.bp_v2,
+        'restrict_policy':
+            _restrict_policy,
+        'init_capacity':
+            self.params.init_size,
+        'partitioner':
+            self.params.partition_fn,
+        'kv_creator':
+            self.params.kv_creator if self.keep_distribution else None,
+        'max_norm':
+            _max_norm,
+        'distribute_strategy':
+            self.distribute_strategy,
+    }
+    return config
+
+
+class UnshapedEmbedding(Embedding):
+
+  def call(self, ids):
+    ids = tf.convert_to_tensor(ids)
+    ids, _ = tf.unique(ids)
+    return de.shadow_ops.embedding_lookup(self.shadow, ids)
+
+
+class SlotEmbedding(Embedding):
```

Comment:
`Slot` maybe not common word for 'feature group',  *field* maybe better.

Reply:
Accept